### PR TITLE
Adding workflow to push a release image to GHCR

### PR DESCRIPTION
**Pull request by bettinaheim**
This PR takes care of prerequisite 3 of https://github.com/NVIDIA/cuda-quantum/issues/41.

Specifically, his PR
- enables the deployment of the devdeps.ext image to GHCR after pushes to main
- deploys a cuda-quantum image to GHCR that is build using the devdeps.ext image and includes the cuquantum backend*
- updates the llvm-only build to use clang to bootstrap it (which is faster than gcc)
- enables image deployments upon pushes to main and release branches, while build cache updates are pushed after pr merges
- contains some clean up and improvements of the workflows.

This PR does not yet include the setup to have the cuda-quantum image include other backends that are not part of this repository.

*) to be confirmed

### CODE REVIEW

File: .github/workflows/build_packages.yml [hunk 1]
```diff
@@ -57,26 +67,50 @@ jobs:
...
+        with:
+          path: ${{ steps.cudaq_image.outputs.tar_archive }}
+          key: ${{ steps.cudaq_image.outputs.tar_cache }}
 
       - name: Validate cuda-quantum image
         run: |
-          docker run --rm -dit --name cuda-quantum cuda-quantum:latest
+          docker run --rm -dit --name cuda-quantum ${{ steps.cudaq_image.outputs.image_name }}:${{ steps.cudaq_image.outputs.image_tag }}
           docker cp scripts/validate_container.sh cuda-quantum:"/home/cudaq/validate_container.sh"
           docker exec -e TERM=xterm cuda-quantum bash validate_container.sh > /tmp/validation.out
```
Comment (schweitzpgi) [inline .github/workflows/build_packages.yml:115]: I think we need to reconsider squirrelling away output into various temporary files exclusively. It's difficult for people not familiar with the details to figure out what happened when things break.

Some possibilities:
- use `tee` to both capture the output and let it be printed in the stdout stream.
- postprocess such that the captured output, if not empty, is concatenated to stdout.

File: .github/workflows/config/spelling_allowlist.txt [hunk 2]
```diff
@@ -0,0 +1,78 @@
+NVIDIA
```
Comment (schweitzpgi) [inline .github/workflows/config/spelling_allowlist.txt:1]: Can this be sorted so it's easier to figure out what's already in the list?

File: Developing.md [hunk 3]
```diff
@@ -1,7 +1,7 @@
 # Getting Started with Developing CUDA Quantum
 
 This document contains guidelines for contributing to the code in this
-repository. This document is relevant primarily for contributions to the NVQ++
+repository. This document is relevant primarily for contributions to the `nvq++`
```
Comment (schweitzpgi) [inline Developing.md:4]: Probably could split the documentation updates into a separate PR. (This one is really long.)

File: docs/sphinx/examples/cpp/other/builder/vqe_h2_builder.cpp [hunk 4]
```diff
@@ -25,7 +27,7 @@ namespace cudaq {
 // Define a function that applies a general SO(4) rotation to
 // the builder on the provided qubits with the provided parameters.
 // Note we keep this qubit and parameter arguments as auto as these
-// will default to taking the qubits and variational parameters (QuakeValues)
+// will default to taking the qubits and variational parameters (`QuakeValue`s)
```
Comment (schweitzpgi) [inline docs/sphinx/examples/cpp/other/builder/vqe_h2_builder.cpp:30]: This isn't a doxygen comment so the backticks don't do anything here. In doxygen, they are the equivalent to inserting `<code> ... </code>` afaict.

**Review by schweitzpgi — COMMENTED** [review]
Whew! That's a lot of changes. I wonder if they could've been split up into a few different chunks.

Glad to get rid of the old spelling checker. :)

File: docs/sphinx/examples/cpp/other/builder/vqe_h2_builder.cpp (hunk 4 above)
Comment (bettinaheim) [reply docs/sphinx/examples/cpp/other/builder/vqe_h2_builder.cpp:30]: The spell checker is configured to ignore anything marked as code (marked by backticks). The spell checker checks for spelling mistakes in comments since it seemed a reasonable option - we use line comments for doc comments as well, and we use normal comments in samples that we should spell check. Plus, I think it is easier to just write all comments similar to doc comments to build the habit, rather than omitting the ticks in comment but not in doc comments. 

In short, the backticks are added so the spell checker does not complain about QuakeValue (which refers to code and hence indicated as such).

File: Developing.md (hunk 3 above)
Comment (bettinaheim) [reply Developing.md:4]: Yes, the spell checker update go bigger than expected - I didn't per se plan that but was annoyed enough with it to want to set it up properly. I can pull the repo checks changes into their own PR.

File: .github/workflows/build_packages.yml (hunk 1 above)
Comment (bettinaheim) [reply .github/workflows/build_packages.yml:115]: I am not sure I follow what you suggest here; are you asking if we can show the output of validate_container.sh in the log with the rest of what the step does rather than only in the step summary?

File: .github/workflows/config/spelling_allowlist.txt (hunk 2 above)
Comment (bettinaheim) [reply .github/workflows/config/spelling_allowlist.txt]: Moved to a separate PR, but yes.

**Comment by bettinaheim** [comment]
@schweitzpgi I split out the changes to the repo checks into their own PR to keep this PR more to what was originally intended.

**Review by schweitzpgi — APPROVED** [review]
Thanks, Bettina.
